polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -789,362 +659,334 @@ impl FragmentTree {
+	fn populate_chain(&mut self, storage: &CandidateStorage) {
+		let mut cumulative_modifications = if let Some(last_candidate) = self.chain.last() {
+			last_candidate.cumulative_modifications.clone()
+		} else {
+			ConstraintModifications::identity()
+		};
+		let Some(mut earliest_rp) = self.earliest_relay_parent() else { return };
 
 		loop {
-			let sweep_start = self.nodes.len();
-
-			if Some(sweep_start) == last_sweep_start {
-				break
+			if self.chain.len() > self.scope.max_depth {
+				break;
 			}
 
-			let parents: Vec<NodePointer> = if let Some(last_start) = last_sweep_start {
-				(last_start..self.nodes.len()).map(NodePointer::Storage).collect()
-			} else {
-				initial_bases.clone()
-			};
+			let child_constraints =
+				match self.scope.base_constraints.apply_modifications(&cumulative_modifications) {
+					Err(e) => {
+						gum::debug!(
+							target: LOG_TARGET,
+							new_parent_head = ?cumulative_modifications.required_parent,
+							err = ?e,
+							"Failed to apply modifications",
+						);
 
-			// 1. get parent head and find constraints
-			// 2. iterate all candidates building on the right head and viable relay parent
-			// 3. add new node
-			for parent_pointer in parents {
-				let (modifications, child_depth, earliest_rp) = match parent_pointer {
-					NodePointer::Root =>
-						(ConstraintModifications::identity(), 0, self.scope.earliest_relay_parent()),
-					NodePointer::Storage(ptr) => {
-						let node = &self.nodes[ptr];
-						let parent_rp = self
-							.scope
-							.ancestor_by_hash(&node.relay_parent())
-							.or_else(|| {
-								// if the relay-parent is out of scope _and_ it is in the tree,
-								// it must be a candidate pending availability.
-								self.scope
-									.get_pending_availability(&node.candidate_hash)
-									.map(|c| c.relay_parent.clone())
-							})
-							.expect("All nodes in tree are either pending availability or within scope; qed");
-
-						(node.cumulative_modifications.clone(), node.depth + 1, parent_rp)
+						break
 					},
+					Ok(c) => c,
 				};
 
-				if child_depth > self.scope.max_depth {
+			let required_head_hash = child_constraints.required_parent.hash();
+			// Even though we don't allow parachain forks under the same active leaf, they may still
+			// appear under different relay chain forks, hence the iterator below.
+			let possible_children = storage.possible_para_children(&required_head_hash);
+			let mut added_child = false;
+			for candidate in possible_children {
+				// Add one node to chain if
+				// 1. it does not introduce a fork or a cycle.
+				// 2. parent hash is correct.
+				// 3. relay-parent does not move backwards.
+				// 4. all non-pending-availability candidates have relay-parent in scope.
+				// 5. candidate outputs fulfill constraints
+
+				if self.is_fork_or_cycle(
+					candidate.parent_head_data_hash(),
+					Some(candidate.output_head_data_hash()),
+				) {
 					continue
```

Comment:
well it can happen if there's a parachain fork or cycle on another relay chain fork. the candidate storage holds all candidates across all relay parents. so when populating the chain, we may attempt to introduce a candidate that is present on another relay chain fork (but those candidates will probably not pass these sanity checks). but it may not neccesarily mean something's wrong